core/src/main/java/org/elasticsearch/index/query/TermsQueryBuilder.java

```diff
@@ -207,4 +289,130 @@ public void doXContent(XContentBuilder builder, Params params) throws IOExceptio
+            terms.addAll(extractedValues);
+        }
+    }
+
+    static Query handleTermsQuery(List<Object> terms, String fieldName, QueryShardContext context, boolean disableCoord) {
+        MappedFieldType fieldType = context.fieldMapper(fieldName);
+        String indexFieldName;
+        if (fieldType != null) {
+            indexFieldName = fieldType.names().indexName();
+        } else {
+            indexFieldName = fieldName;
+        }
+
+        Query query;
+        if (context.isFilter()) {  // NO COMMIT: not exactly sure how to handle this case, right now using a filter flag
+            if (fieldType != null) {
+                query = fieldType.termsQuery(terms, context);
+            } else {
+                BytesRef[] filterValues = new BytesRef[terms.size()];
+                for (int i = 0; i < filterValues.length; i++) {
+                    filterValues[i] = BytesRefs.toBytesRef(terms.get(i));
+                }
+                query = new TermsQuery(indexFieldName, filterValues);
+            }
+        } else {
+            BooleanQuery bq = new BooleanQuery(disableCoord);
+            for (Object term : terms) {
+                if (fieldType != null) {
+                    bq.add(fieldType.termQuery(term, context), BooleanClause.Occur.SHOULD);
+                } else {
+                    bq.add(new TermQuery(new Term(indexFieldName, BytesRefs.toBytesRef(term))), BooleanClause.Occur.SHOULD);
+                }
+            }
+            query = bq;
+        }
+        return query;
+    }
+
+    @Override
+    public QueryValidationException validate() {
+        QueryValidationException validationException = null;
+        if (this.fieldName == null) {
+            validationException = addValidationError("field name cannot be null.", validationException);
+        }
+        if (isTermsLookupQuery()) {
+            validationException = validateTermsLookup(validationException);
+        }
+        return validationException;
+    }
+
+    private QueryValidationException validateTermsLookup(QueryValidationException validationException) {
+        if (lookupType == null) {
+            validationException = addValidationError("[terms] query lookup element requires specifying the type.", validationException);
+        }
+        if (lookupId == null) {
+            validationException = addValidationError("[terms] query lookup element requires specifying the id.", validationException);
+        }
+        if (lookupPath == null) {
+            validationException = addValidationError("[terms] query lookup element requires specifying the path.", validationException);
+        }
+        return validationException;
+    }
+
+    @Override
+    protected TermsQueryBuilder doReadFrom(StreamInput in) throws IOException {
+        TermsQueryBuilder termsQueryBuilder = new TermsQueryBuilder(in.readString(), in.readGenericValue());
+        if (termsQueryBuilder.isTermsLookupQuery()) {
```

Comment:
do we decide what to do based on fields that we haven't read yet?  I think this conditional will always be false. Which also means that we are not testing this, which we should.